types/react-redux/index.d.ts

```diff
@@ -445,3 +447,240 @@ export const ReactReduxContext: Context<ReactReduxContextValue>;
  * multiple actions dispatched outside of React only result in a single render update.
  */
 export function batch(cb: () => void): void;
+
+// tslint:disable:no-unnecessary-generics
+
+/**
+ * A hook to bind action creators to the redux store's `dispatch` function
+ * similar to how redux's `bindActionCreators` works.
+ *
+ * Supports passing a single action creator, an array/tuple of action
+ * creators, or an object of action creators.
+ *
+ * Any arguments passed to the created callbacks are passed through to
+ * your functions.
+ *
+ * This hook takes a dependencies array as an optional second argument,
+ * which when passed ensures referential stability of the created callbacks.
+ *
+ * @param actions the action creators to bind
+ * @param deps (optional) dependencies array to control referential stability
+ *
+ * @returns callback(s) bound to store's `dispatch` function
+ *
+ * @example
+ *
+ * import React from 'react'
+ * import { useActions } from 'react-redux'
+ *
+ * const increaseCounter = ({ amount }) => ({
+ *   type: 'increase-counter',
+ *   amount,
+ * })
+ *
+ * export const CounterComponent = ({ value }) => {
+ *   // supports passing an object of action creators
+ *   const { increaseCounterByOne, increaseCounterByTwo } = useActions({
+ *     increaseCounterByOne: () => increaseCounter(1),
+ *     increaseCounterByTwo: () => increaseCounter(2),
+ *   }, [])
+ *
+ *   // supports passing an array/tuple of action creators
+ *   const [increaseCounterByThree, increaseCounterByFour] = useActions([
+ *     () => increaseCounter(3),
+ *     () => increaseCounter(4),
+ *   ], [])
+ *
+ *   // supports passing a single action creator
+ *   const increaseCounterBy5 = useActions(() => increaseCounter(5), [])
+ *
+ *   // passes through any arguments to the callback
+ *   const increaseCounterByX = useActions(x => increaseCounter(x), [])
+ *
+ *   return (
+ *     <div>
+ *       <span>{value}</span>
+ *       <button onClick={increaseCounterByOne}>Increase counter by 1</button>
+ *     </div>
+ *   )
+ * }
+ */
+export function useActions<
+    T extends [ActionCreator<A>, ActionCreator<A>, ActionCreator<A>, ActionCreator<A>, ActionCreator<A>, ActionCreator<A>, ActionCreator<A>, ActionCreator<A>, ActionCreator<A>],
+    A = AnyAction
```

Comment:
I think there is no need for the `A = AnyAction`. Actually, this generic type looks useless. Also, it causes problems with some Redux middlewares like redux-thunk.

Reply:
The reason I did it this way is to be similar to how `bindActionCreators` from redux is [typed](https://github.com/reduxjs/redux/blob/da3d89fc229843c6514d29fbb204fc073359fd54/index.d.ts#L469). I don't think it hurts to have that extra generic parameter since in pretty much all cases you are not going to specify any of the parameters anyways. However, as you say, it can easily be replaced with `any` or `AnyAction`, so if you still think I should remove it, I will do so.

Reply:
Let's have a minimal implementation for it without extra generics. Also, it is important to don't restrict action creator return type to `AnyAction` for compatibility with thunks or other similar middlewares.

Reply:
Please see [MrWolfZ/DefinitelyTyped#1](https://github.com/MrWolfZ/DefinitelyTyped/pull/1).

Reply:
I have merged the PR.

Reply:
I think we need feedback from more people on this. Redux bound actions are confusing to type (I would say nearly impossible) since the return value is determined by the middleware you're running. We can't solve every case of this, but in the most basic case of this, setting this to `any` will cause a lot of errors since we can't do conditional checking between action creators (which return an object) and thunks (which return whatever the thunk returns).

The existing typings have built-in support for `redux-thunk`, for instance, which is missing from these types:
```
TActionCreator extends (
    ...args: infer TParams
  ) => (...args: any[]) => infer TReturn
    ? (...args: TParams) => TReturn
    : TActionCreator
```

@surgeboris  and @rhys-vdw had to resolve a middleware-related bug/conflict in https://github.com/DefinitelyTyped/DefinitelyTyped/issues/30740, so they may have a specific interest here.

Reply:
I'll work on a PR tonight to allow any function as an action, but still support thunk return values. That should cover the simplest cases. It shouldn't be anywhere near as difficult as `connect()` has been.

I'll also add some tests so we're not guessing anymore.

Reply:
Hey @threehams, just responding to the mention. Don't think I've got anything of value to add here as I haven't really dug into hooks yet, but thanks for the heads up. I'll keep an eye on the issue nonetheless.